docs/design/job_architecture.md

```diff
@@ -482,44 +306,187 @@ All cases:
-4. `JobComm` does:
+* The `JobManager` runs `check_workspace_jobs` on EE2 with the current workspace ID to fetch the list of jobs associated with the workspace. This is filtered to include only those in `cell_list`, and then `Job` objects are built from the data.
+* `JobComm` does:
   * Starts the lookup loop thread.
   * On the first pass, this looks up status of all jobs and pushes them forward to the browser.
   * If any jobs are in a terminal state, they'll stopped being looked up automatically. If all jobs are terminated, then the loop thread itself stops.
 
-## JobComm status lookup loop.
+
+### Running an app
+
+These steps define the process by which jobs are created by running an app.
+
+* User clicks "Run" on app or batch cell in the browser.
+  * Cell provides app_id, cell_id, run_id, and parameters.
+  * Invokes `biokbase.narrative.appmanager.AppManager.run_app` or `run_app_batch`.
+* `AppManager.run_app` validates the following bits of information before passing them on to EE2:
+  * App (based on id, version, and spec).
+  * Parameters (based on the app spec).
+* `AppManager.run_app` preparation and start:
+  * Convert app params from user-readable to machine-understandable (via the spec input_mapping).
+  * Fetch cell id, run id, workspace id, user token.
+  * Create an agent token on behalf of the user. This effectively makes a new authentication token that has a two-week lifetime, separate from the current login token. For example, if the user's current login token has a remaining lifespan of 1 hour, then the new job will be able to continue long past that.
+  * Submit all of the above to the Execution Engine (EE2) endpoint `run_job` or `run_job_batch`.
+* Get response from EE2
+* Send a `RUN_STATUS` message to the browser, giving the results of the job submission; this will either contain a job ID or list of job IDs, if the submission was successful, or an error if not.
+* Assuming the submission was successful, the job IDs are combine with info from the job submission to create `biokbase.narrative.jobs.job.Job` objects. These are registered with the `biokbase.narrative.jobs.jobmanager.JobManager` singleton object.
+* On the frontend, the cell receiving the `RUN_STATUS` message will send `STATUS` requests to the backend to track the progress of the job. Other message types (`LOGS`, `CANCEL`, `RETRY`, etc.) are sent as needed.
+  * Batch cells have an update mechanism that triggers a `STATUS` request a short period after receiving the last update from the backend.
+  * App cells use the job lookup loop to keep up-to-date with job status.
+
+
+### App initialisation on reload
+
+These steps are taken when an app with a running (or previously-run) job starts up.
+* Job ID(s) are retrieved from the app meta data
+* If the job was not in a terminal state when it was saved, the app requests `STATUS` from the backend.

```

Comment:
How to retries work if the narrative didn't get a chance to save itself after submitting the retry request? Presumably the retried job will be terminal and the new job won't be in the metadata

Reply:
Added clarification here

Reply:
:+1: